File: Chapter3/Codes_Figures/jorda_lp_figure.R
# ===========================================================================
# Jordà (2005) Local Projections Example with Real U.S. Data — R Version
# ===========================================================================
# Comparing LP vs VAR impulse responses for monetary policy analysis
# Using FRED data: GDP (GDPC1), GDP Deflator (GDPDEF), Federal Funds Rate
#
# Instructions:
#   1. Place GDPC1.xlsx, GDPDEF.xlsx, and FEDFUNDS.xlsx in your working
#      directory (or adjust the paths below).
#   2. Install required packages if needed (see below).
#   3. Source this script: source("jorda_lp_real_data.R")
#. Author: Alessia Paccagnini
#. Textbook:Macroeconometrics
# ===========================================================================

# --- Install packages (uncomment if needed) --------------------------------
# install.packages(c("readxl", "vars", "sandwich", "lmtest", "mFilter"))

library(readxl)
library(vars)
library(sandwich)
library(lmtest)
library(mFilter)

# ===========================================================================
# 1. Load Data
# ===========================================================================
cat("Loading FRED data...\n")

# --- Set file paths (adjust if your files are elsewhere) -------------------
# For Google Colab / Jupyter-R you can upload files first, then point here.
gdp_file      <- "GDPC1.xlsx"
deflator_file <- "GDPDEF.xlsx"
fedfunds_file <- "FEDFUNDS.xlsx"

gdp_raw      <- read_excel(gdp_file,      sheet = "Quarterly")
deflator_raw <- read_excel(deflator_file,  sheet = "Quarterly")
fedfunds_raw <- read_excel(fedfunds_file,  sheet = "Monthly")

cat(sprintf("  GDP:       %d obs\n", nrow(gdp_raw)))
cat(sprintf("  Deflator:  %d obs\n", nrow(deflator_raw)))
cat(sprintf("  Fed Funds: %d obs\n", nrow(fedfunds_raw)))

# ===========================================================================
# 2. Process and Merge
# ===========================================================================

# Rename columns
colnames(gdp_raw)      <- c("date", "gdp")
colnames(deflator_raw) <- c("date", "deflator")
colnames(fedfunds_raw) <- c("date", "fedfunds")

# Convert dates
gdp_raw$date      <- as.Date(gdp_raw$date)
deflator_raw$date <- as.Date(deflator_raw$date)
fedfunds_raw$date <- as.Date(fedfunds_raw$date)

# Fed funds: monthly → quarterly average
fedfunds_raw$quarter <- as.Date(
  paste0(format(fedfunds_raw$date, "%Y"),
         "-", sprintf("%02d", (as.numeric(format(fedfunds_raw$date, "%m")) - 1) %/% 3 * 3 + 1),
         "-01"))
fedfunds_q <- aggregate(fedfunds ~ quarter, data = fedfunds_raw, FUN = mean)
colnames(fedfunds_q) <- c("date", "fedfunds")

# Align GDP and deflator to quarter-start dates
quarter_start <- function(d) {
  m <- as.numeric(format(d, "%m"))
  q_start <- (m - 1) %/% 3 * 3 + 1
  as.Date(paste0(format(d, "%Y"), "-", sprintf("%02d", q_start), "-01"))
}
gdp_raw$date      <- quarter_start(gdp_raw$date)
deflator_raw$date <- quarter_start(deflator_raw$date)

# Merge
df <- merge(gdp_raw, deflator_raw, by = "date")
df <- merge(df, fedfunds_q, by = "date")
df <- df[order(df$date), ]

cat(sprintf("\nMerged data: %d observations\n", nrow(df)))
cat(sprintf("Date range:  %s to %s\n", min(df$date), max(df$date)))

# ===========================================================================
# 3. Construct Variables
# ===========================================================================

# Output gap via HP filter (lambda = 1600)
df$log_gdp <- 100 * log(df$gdp)
hp <- hpfilter(ts(df$log_gdp, frequency = 4), freq = 1600)
df$output_gap <- as.numeric(hp$cycle)

# Inflation: 4-quarter change in log GDP deflator (annualised)
df$log_deflator <- 100 * log(df$deflator)
df$inflation <- c(rep(NA, 4), diff(df$log_deflator, lag = 4))

# Federal funds rate
df$fed_funds <- df$fedfunds

# Sample: 1960Q1 – 2007Q4
sample_idx <- df$date >= as.Date("1960-01-01") & df$date <= as.Date("2007-12-31")
df_sample  <- df[sample_idx, c("date", "output_gap", "inflation", "fed_funds")]
df_sample  <- na.omit(df_sample)

start_q <- sprintf("%sQ%d", format(min(df_sample$date), "%Y"),
                   (as.numeric(format(min(df_sample$date), "%m")) - 1) %/% 3 + 1)
end_q   <- sprintf("%sQ%d", format(max(df_sample$date), "%Y"),
                   (as.numeric(format(max(df_sample$date), "%m")) - 1) %/% 3 + 1)

cat(sprintf("\nEstimation sample: %s – %s  (T = %d)\n", start_q, end_q, nrow(df_sample)))
cat("\nDescriptive statistics:\n")
print(summary(df_sample[, c("output_gap", "inflation", "fed_funds")]))

# ===========================================================================
# 4. Estimate VAR Models
# ===========================================================================
cat("\n", strrep("=", 60), "\n")
cat("Estimating VAR models...\n")
cat(strrep("=", 60), "\n")

Y <- df_sample[, c("output_gap", "inflation", "fed_funds")]

var4 <- VAR(Y, p = 4, type = "const")
var1 <- VAR(Y, p = 1, type = "const")

cat(sprintf("  VAR(4)  AIC = %.2f   BIC = %.2f\n",
            AIC(var4), BIC(var4)))
cat(sprintf("  VAR(1)  AIC = %.2f   BIC = %.2f\n",
            AIC(var1), BIC(var1)))

# Cholesky IRFs (ordering: output_gap → inflation → fed_funds)
H <- 20
irf_var4 <- irf(var4, impulse = "fed_funds", n.ahead = H,
                 ortho = TRUE, ci = 0.95, boot = TRUE, runs = 500)
irf_var1 <- irf(var1, impulse = "fed_funds", n.ahead = H,
                 ortho = TRUE, ci = 0.95, boot = TRUE, runs = 500)

# ===========================================================================
# 5. Estimate Local Projections
# ===========================================================================
cat("\n", strrep("=", 60), "\n")
cat("Estimating Local Projections...\n")
cat(strrep("=", 60), "\n")

estimate_lp <- function(data, shock_var, response_var, H, n_lags) {
  # Jordà (2005) LP with Newey–West standard errors
  #
  # Args:
  #   data         : data.frame with variables (no date column)
  #   shock_var    : character, name of shock variable
  #   response_var : character, name of response variable
  #   H            : integer, maximum horizon
  #   n_lags       : integer, number of control lags
  #
  # Returns:
  #   list(irf, se, ci_lower, ci_upper) — each a numeric vector of length H+1

  vars_names <- colnames(data)
  n <- nrow(data)

  irf_vec <- numeric(H + 1)
  se_vec  <- numeric(H + 1)

  for (h in 0:H) {
    # Dependent variable: y_{t+h}
    y <- c(data[(1 + h):n, response_var], rep(NA, h))

    # Controls: lags 1..n_lags of every variable
    X <- data[, shock_var]  # contemporaneous shock
    for (v in vars_names) {
      for (lag in 1:n_lags) {
        lagged <- c(rep(NA, lag), data[1:(n - lag), v])
        X <- cbind(X, lagged)
      }
    }
    X <- cbind(1, X)  # add constant

    # Drop incomplete rows
    ok <- complete.cases(cbind(y, X))
    y_clean <- y[ok]
    X_clean <- as.matrix(X[ok, ])

    # OLS
    fit <- lm(y_clean ~ X_clean - 1)  # -1 because constant already in X

    # Newey–West standard errors
    nw_lags <- max(h + 1, 4)
    vcov_nw <- NeweyWest(fit, lag = nw_lags, prewhite = FALSE)

    # Coefficient on shock (column 2: after constant)
    irf_vec[h + 1] <- coef(fit)[2]
    se_vec[h + 1]  <- sqrt(vcov_nw[2, 2])
  }

  list(irf      = irf_vec,
       se       = se_vec,
       ci_lower = irf_vec - 1.96 * se_vec,
       ci_upper = irf_vec + 1.96 * se_vec)
}

Y_mat <- df_sample[, c("output_gap", "inflation", "fed_funds")]

cat("  Output gap response...\n")
lp_y  <- estimate_lp(Y_mat, "fed_funds", "output_gap", H, n_lags = 4)

cat("  Inflation response...\n")
lp_pi <- estimate_lp(Y_mat, "fed_funds", "inflation",  H, n_lags = 4)

cat("  Fed funds own response...\n")
lp_ff <- estimate_lp(Y_mat, "fed_funds", "fed_funds",  H, n_lags = 4)

cat("  Done!\n")

# ===========================================================================
# 6. Publication-Quality Figure
# ===========================================================================
cat("\n", strrep("=", 60), "\n")
cat("Creating figure...\n")
cat(strrep("=", 60), "\n")

# Extract VAR(4) IRFs
var4_y   <- irf_var4$irf$fed_funds[, "output_gap"]
var4_lo_y <- irf_var4$Lower$fed_funds[, "output_gap"]
var4_hi_y <- irf_var4$Upper$fed_funds[, "output_gap"]

var4_pi    <- irf_var4$irf$fed_funds[, "inflation"]
var4_lo_pi <- irf_var4$Lower$fed_funds[, "inflation"]
var4_hi_pi <- irf_var4$Upper$fed_funds[, "inflation"]

# VAR(1) IRFs
var1_y  <- irf_var1$irf$fed_funds[, "output_gap"]
var1_pi <- irf_var1$irf$fed_funds[, "inflation"]

horizons <- 0:H

# Colours
col_var4 <- "#2166AC"
col_lp   <- "#B2182B"
col_var1 <- "#4DAF4A"

# Save to file
png("jorda_lp_example_real_data.png", width = 14, height = 5.5,
    units = "in", res = 200)
par(mfrow = c(1, 2), mar = c(4.5, 4.5, 3, 1), family = "serif", cex = 1.05)

# --- Panel (a): Output Gap -------------------------------------------------
ym <- max(abs(c(var4_lo_y, var4_hi_y, lp_y$ci_lower, lp_y$ci_upper,
                var1_y))) * 1.15

plot(NA, xlim = c(0, H), ylim = c(-ym, ym),
     xlab = "Quarters after shock", ylab = "Percent",
     main = "(a) Response of Output Gap to Fed Funds Shock",
     xaxt = "n", font.main = 2)
axis(1, at = seq(0, 20, 4))
abline(h = 0, col = "black", lwd = 1)

# VAR(4) CI
polygon(c(horizons, rev(horizons)),
        c(var4_lo_y, rev(var4_hi_y)),
        col = adjustcolor(col_var4, alpha.f = 0.15), border = NA)
lines(horizons, var4_y, col = col_var4, lwd = 2.5)

# LP CI
polygon(c(horizons, rev(horizons)),
        c(lp_y$ci_lower, rev(lp_y$ci_upper)),
        col = adjustcolor(col_lp, alpha.f = 0.15), border = NA)
lines(horizons, lp_y$irf, col = col_lp, lwd = 2.5, lty = 2)

# VAR(1)
lines(horizons, var1_y, col = col_var1, lwd = 2, lty = 3)

legend("bottomright",
       legend = c("VAR(4)", "Local Projections", expression(paste("VAR(1) — misspecified"))),
       col = c(col_var4, col_lp, col_var1),
       lwd = c(2.5, 2.5, 2), lty = c(1, 2, 3),
       bg = "white", cex = 0.95)

# --- Panel (b): Inflation ---------------------------------------------------
ym_pi <- max(abs(c(var4_lo_pi, var4_hi_pi, lp_pi$ci_lower, lp_pi$ci_upper,
                    var1_pi))) * 1.15

plot(NA, xlim = c(0, H), ylim = c(-ym_pi, ym_pi),
     xlab = "Quarters after shock", ylab = "Percent",
     main = "(b) Response of Inflation to Fed Funds Shock",
     xaxt = "n", font.main = 2)
axis(1, at = seq(0, 20, 4))
abline(h = 0, col = "black", lwd = 1)

# VAR(4) CI
polygon(c(horizons, rev(horizons)),
        c(var4_lo_pi, rev(var4_hi_pi)),
        col = adjustcolor(col_var4, alpha.f = 0.15), border = NA)
lines(horizons, var4_pi, col = col_var4, lwd = 2.5)

# LP CI
polygon(c(horizons, rev(horizons)),
        c(lp_pi$ci_lower, rev(lp_pi$ci_upper)),
        col = adjustcolor(col_lp, alpha.f = 0.15), border = NA)
lines(horizons, lp_pi$irf, col = col_lp, lwd = 2.5, lty = 2)

# VAR(1)
lines(horizons, var1_pi, col = col_var1, lwd = 2, lty = 3)

legend("topright",
       legend = c("VAR(4)", "Local Projections", expression(paste("VAR(1) — misspecified"))),
       col = c(col_var4, col_lp, col_var1),
       lwd = c(2.5, 2.5, 2), lty = c(1, 2, 3),
       bg = "white", cex = 0.95)

# Main title
mtext(paste0("Impulse Responses to a Monetary Policy Shock: LP vs VAR\n",
             "U.S. Quarterly Data, ", start_q, "–", end_q),
      side = 3, outer = TRUE, line = -2, cex = 1.2, font = 2)

dev.off()
cat("Figure saved as 'jorda_lp_example_real_data.png'\n")

# ===========================================================================
# 7. Summary
# ===========================================================================
cat("\n", strrep("=", 60), "\n")
cat("SUMMARY: LP vs VAR Comparison (Real U.S. Data)\n")
cat(strrep("=", 60), "\n")
cat(sprintf("\nSample: %s – %s  (T = %d)\n", start_q, end_q, nrow(df_sample)))
cat("Variables: Output Gap, Inflation, Federal Funds Rate\n")
cat("Identification: Cholesky (output → inflation → fed funds)\n")

cat(sprintf("\n--- Output Gap Response (h = 8) ---\n"))
cat(sprintf("  VAR(4): %7.4f  [%7.4f, %7.4f]\n", var4_y[9], var4_lo_y[9], var4_hi_y[9]))
cat(sprintf("  LP:     %7.4f  [%7.4f, %7.4f]\n", lp_y$irf[9], lp_y$ci_lower[9], lp_y$ci_upper[9]))
cat(sprintf("  VAR(1): %7.4f  (misspecified)\n", var1_y[9]))

cat(sprintf("\n--- Inflation Response (h = 8) ---\n"))
cat(sprintf("  VAR(4): %7.4f  [%7.4f, %7.4f]\n", var4_pi[9], var4_lo_pi[9], var4_hi_pi[9]))
cat(sprintf("  LP:     %7.4f  [%7.4f, %7.4f]\n", lp_pi$irf[9], lp_pi$ci_lower[9], lp_pi$ci_upper[9]))
cat(sprintf("  VAR(1): %7.4f  (misspecified)\n", var1_pi[9]))

cat("\n", strrep("=", 60), "\n")

File: Chapter3/Codes_Figures/jorda_lp_figure.py
"""
Jordà (2005) Local Projections Example with Real U.S. Data
==========================================================
Comparing LP vs VAR impulse responses for monetary policy analysis
Using FRED data: GDP, GDP Deflator, Federal Funds Rate

Google Colab version — upload Excel files when prompted.
"""

# ============================================================
# 0. Install dependencies (Colab)
# ============================================================
# !pip install statsmodels openpyxl

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from statsmodels.tsa.api import VAR
from statsmodels.tsa.filters.hp_filter import hpfilter
from statsmodels.regression.linear_model import OLS
from statsmodels.tools import add_constant
import warnings
warnings.filterwarnings('ignore')

# Set style for publication-quality figures
plt.style.use('seaborn-v0_8-whitegrid')
plt.rcParams['font.family'] = 'serif'
plt.rcParams['font.size'] = 11
plt.rcParams['axes.linewidth'] = 1.2
plt.rcParams['axes.edgecolor'] = 'black'

# ============================================================
# 1. Upload and Load FRED Data (Colab file upload)
# ============================================================
print("=" * 60)
print("Please upload the three FRED Excel files when prompted:")
print("  1) GDPC1.xlsx   (Real GDP)")
print("  2) GDPDEF.xlsx  (GDP Deflator)")
print("  3) FEDFUNDS.xlsx (Federal Funds Rate)")
print("=" * 60)

from google.colab import files

print("\n>>> Upload GDPC1.xlsx (Real GDP):")
uploaded = files.upload()
gdp_file = list(uploaded.keys())[0]

print("\n>>> Upload GDPDEF.xlsx (GDP Deflator):")
uploaded = files.upload()
deflator_file = list(uploaded.keys())[0]

print("\n>>> Upload FEDFUNDS.xlsx (Federal Funds Rate):")
uploaded = files.upload()
fedfunds_file = list(uploaded.keys())[0]

# ============================================================
# 2. Read and Process Data
# ============================================================
print("\nLoading data...")

gdp = pd.read_excel(gdp_file, sheet_name='Quarterly')
deflator = pd.read_excel(deflator_file, sheet_name='Quarterly')

# Fed funds is monthly
try:
    fedfunds = pd.read_excel(fedfunds_file, sheet_name='Monthly')
except Exception:
    fedfunds = pd.read_excel(fedfunds_file, sheet_name='Quarterly')

print(f"  GDP shape:       {gdp.shape}")
print(f"  Deflator shape:  {deflator.shape}")
print(f"  Fed Funds shape: {fedfunds.shape}")

# Rename columns (FRED format: observation_date, SERIES_NAME)
gdp.columns = ['date', 'gdp']
deflator.columns = ['date', 'deflator']
fedfunds.columns = ['date', 'fedfunds']

# Convert dates
gdp['date'] = pd.to_datetime(gdp['date'])
deflator['date'] = pd.to_datetime(deflator['date'])
fedfunds['date'] = pd.to_datetime(fedfunds['date'])

# Fed funds: monthly → quarterly (average)
fedfunds['quarter'] = fedfunds['date'].dt.to_period('Q')
fedfunds_q = fedfunds.groupby('quarter')['fedfunds'].mean().reset_index()
fedfunds_q['date'] = fedfunds_q['quarter'].dt.to_timestamp()
fedfunds_q = fedfunds_q[['date', 'fedfunds']]

# Align GDP and deflator to quarter start
gdp['date'] = gdp['date'].dt.to_period('Q').dt.to_timestamp()
deflator['date'] = deflator['date'].dt.to_period('Q').dt.to_timestamp()

# Merge
df = gdp.merge(deflator, on='date', how='inner')
df = df.merge(fedfunds_q, on='date', how='inner')
df = df.set_index('date').sort_index()

print(f"\nMerged data: {df.shape[0]} observations")
print(f"Date range:  {df.index[0].strftime('%Y-%m')} to {df.index[-1].strftime('%Y-%m')}")

# ============================================================
# 3. Construct Variables
# ============================================================

# Output gap via HP filter (λ = 1600 for quarterly data)
df['log_gdp'] = 100 * np.log(df['gdp'])
cycle, trend = hpfilter(df['log_gdp'].dropna(), lamb=1600)
df['output_gap'] = cycle

# Inflation: annualised 4-quarter change in log GDP deflator
df['log_deflator'] = 100 * np.log(df['deflator'])
df['inflation'] = df['log_deflator'].diff(4)

# Federal funds rate (already in percent)
df['fed_funds'] = df['fedfunds']

# Sample: 1960Q1 – 2007Q4 (pre-crisis, avoids ZLB)
df_sample = df.loc['1960-01-01':'2007-12-31',
                    ['output_gap', 'inflation', 'fed_funds']].dropna()

start_q = f"{df_sample.index[0].year}Q{(df_sample.index[0].month - 1) // 3 + 1}"
end_q   = f"{df_sample.index[-1].year}Q{(df_sample.index[-1].month - 1) // 3 + 1}"

print(f"\nEstimation sample: {start_q} – {end_q}  (T = {len(df_sample)})")
print("\nDescriptive statistics:")
print(df_sample.describe().round(2))

# ============================================================
# 4. Estimate VAR Models
# ============================================================
print("\n" + "=" * 60)
print("Estimating VAR models...")
print("=" * 60)

model_var = VAR(df_sample)

# VAR(4) — standard for quarterly data
results_var4 = model_var.fit(4)
print(f"\n  VAR(4)  AIC = {results_var4.aic:.2f}   BIC = {results_var4.bic:.2f}")

# VAR(1) — deliberately misspecified for comparison
results_var1 = model_var.fit(1)
print(f"  VAR(1)  AIC = {results_var1.aic:.2f}   BIC = {results_var1.bic:.2f}")

# Cholesky IRFs (ordering: output_gap → inflation → fed_funds)
H = 20  # 20-quarter horizon
irf_var4 = results_var4.irf(H)
irf_var1 = results_var1.irf(H)

# ============================================================
# 5. Estimate Local Projections
# ============================================================
print("\n" + "=" * 60)
print("Estimating Local Projections...")
print("=" * 60)


def estimate_lp(data, shock_var, response_var, H, n_lags):
    """
    Jordà (2005) Local Projections with Newey–West standard errors.

    Parameters
    ----------
    data : DataFrame — panel of variables
    shock_var : str — name of the shock variable
    response_var : str — name of the response variable
    H : int — maximum horizon
    n_lags : int — number of control lags

    Returns
    -------
    irfs, ses, ci_lower, ci_upper : arrays of length H+1
    """
    irfs = np.zeros(H + 1)
    ses  = np.zeros(H + 1)

    for h in range(H + 1):
        # Dependent variable: y_{t+h}
        y = data[response_var].shift(-h)

        # Controls: lags 1..n_lags of every variable
        X_list = []
        for var in data.columns:
            for lag in range(1, n_lags + 1):
                X_list.append(data[var].shift(lag))

        # Contemporaneous shock + controls + constant
        X_df = pd.concat([data[shock_var]] + X_list, axis=1)
        X_df = add_constant(X_df)

        # Drop NaN rows
        combined = pd.concat([y, X_df], axis=1).dropna()
        y_clean = combined.iloc[:, 0].values
        X_clean = combined.iloc[:, 1:].values

        # OLS with HAC (Newey–West) standard errors
        nw_lags = max(h + 1, 4)
        res = OLS(y_clean, X_clean).fit(
            cov_type='HAC', cov_kwds={'maxlags': nw_lags}
        )

        # Coefficient on the contemporaneous shock (index 1 after constant)
        irfs[h] = res.params[1]
        ses[h]  = res.bse[1]

    ci_lower = irfs - 1.96 * ses
    ci_upper = irfs + 1.96 * ses
    return irfs, ses, ci_lower, ci_upper


n_lags = 4

print("  Output gap response...")
irf_lp_y, se_lp_y, ci_lo_y, ci_hi_y = estimate_lp(
    df_sample, 'fed_funds', 'output_gap', H, n_lags)

print("  Inflation response...")
irf_lp_pi, se_lp_pi, ci_lo_pi, ci_hi_pi = estimate_lp(
    df_sample, 'fed_funds', 'inflation', H, n_lags)

print("  Fed funds own response...")
irf_lp_ff, se_lp_ff, ci_lo_ff, ci_hi_ff = estimate_lp(
    df_sample, 'fed_funds', 'fed_funds', H, n_lags)

print("  Done!")

# ============================================================
# 6. Publication-Quality Figure
# ============================================================
print("\n" + "=" * 60)
print("Creating figure...")
print("=" * 60)

# Cholesky indices (ordering: output_gap=0, inflation=1, fed_funds=2)
shock_idx   = 2
resp_y_idx  = 0
resp_pi_idx = 1

# VAR(4) IRFs and 95 % bands
var4_y   = irf_var4.irfs[:, resp_y_idx,  shock_idx]
var4_pi  = irf_var4.irfs[:, resp_pi_idx, shock_idx]
var4_se_y  = irf_var4.stderr()[:, resp_y_idx,  shock_idx]
var4_se_pi = irf_var4.stderr()[:, resp_pi_idx, shock_idx]
var4_lo_y, var4_hi_y   = var4_y - 1.96 * var4_se_y, var4_y + 1.96 * var4_se_y
var4_lo_pi, var4_hi_pi = var4_pi - 1.96 * var4_se_pi, var4_pi + 1.96 * var4_se_pi

# VAR(1) IRFs (no bands — shown as misspecified benchmark)
var1_y  = irf_var1.irfs[:, resp_y_idx,  shock_idx]
var1_pi = irf_var1.irfs[:, resp_pi_idx, shock_idx]

horizons = np.arange(H + 1)

# Colours
c_var4 = '#2166AC'   # blue
c_lp   = '#B2182B'   # red
c_var1 = '#4DAF4A'   # green
alpha  = 0.15

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# --- Panel (a): Output Gap ---
ax = axes[0]
ax.fill_between(horizons, var4_lo_y, var4_hi_y, color=c_var4, alpha=alpha)
ax.plot(horizons, var4_y, color=c_var4, lw=2.5, label='VAR(4)')
ax.fill_between(horizons, ci_lo_y, ci_hi_y, color=c_lp, alpha=alpha)
ax.plot(horizons, irf_lp_y, color=c_lp, lw=2.5, ls='--', label='Local Projections')
ax.plot(horizons, var1_y, color=c_var1, lw=2, ls=':', label='VAR(1) — misspecified')
ax.axhline(0, color='black', lw=1)
ax.set_xlabel('Quarters after shock', fontsize=13)
ax.set_ylabel('Percent', fontsize=13)
ax.set_title('(a) Response of Output Gap to Fed Funds Shock',
             fontsize=14, fontweight='bold', pad=10)
ax.legend(loc='lower right', fontsize=11, framealpha=0.95,
          edgecolor='gray', fancybox=False)
ax.set_xlim(0, 20)
ax.set_xticks([0, 4, 8, 12, 16, 20])
ax.tick_params(labelsize=11)
ym = max(abs(min(var4_lo_y.min(), ci_lo_y.min(), var1_y.min())),
         abs(max(var4_hi_y.max(), ci_hi_y.max(), var1_y.max()))) * 1.15
ax.set_ylim(-ym, ym)

# --- Panel (b): Inflation ---
ax = axes[1]
ax.fill_between(horizons, var4_lo_pi, var4_hi_pi, color=c_var4, alpha=alpha)
ax.plot(horizons, var4_pi, color=c_var4, lw=2.5, label='VAR(4)')
ax.fill_between(horizons, ci_lo_pi, ci_hi_pi, color=c_lp, alpha=alpha)
ax.plot(horizons, irf_lp_pi, color=c_lp, lw=2.5, ls='--', label='Local Projections')
ax.plot(horizons, var1_pi, color=c_var1, lw=2, ls=':', label='VAR(1) — misspecified')
ax.axhline(0, color='black', lw=1)
ax.set_xlabel('Quarters after shock', fontsize=13)
ax.set_ylabel('Percent', fontsize=13)
ax.set_title('(b) Response of Inflation to Fed Funds Shock',
             fontsize=14, fontweight='bold', pad=10)
ax.legend(loc='upper right', fontsize=11, framealpha=0.95,
          edgecolor='gray', fancybox=False)
ax.set_xlim(0, 20)
ax.set_xticks([0, 4, 8, 12, 16, 20])
ax.tick_params(labelsize=11)
ym_pi = max(abs(min(var4_lo_pi.min(), ci_lo_pi.min(), var1_pi.min())),
            abs(max(var4_hi_pi.max(), ci_hi_pi.max(), var1_pi.max()))) * 1.15
ax.set_ylim(-ym_pi, ym_pi)

fig.suptitle(
    'Impulse Responses to a Monetary Policy Shock: LP vs VAR\n'
    f'U.S. Quarterly Data, {start_q}–{end_q}',
    fontsize=15, fontweight='bold', y=1.02)

plt.tight_layout()
plt.savefig('jorda_lp_example_real_data.png', dpi=200,
            bbox_inches='tight', facecolor='white', edgecolor='none')
plt.show()
print("Figure saved as 'jorda_lp_example_real_data.png'")

# ============================================================
# 7. Summary Statistics
# ============================================================
print("\n" + "=" * 60)
print("SUMMARY: LP vs VAR Comparison (Real U.S. Data)")
print("=" * 60)
print(f"\nSample: {start_q} – {end_q}  (T = {len(df_sample)})")
print("Variables: Output Gap, Inflation, Federal Funds Rate")
print("Identification: Cholesky (output → inflation → fed funds)")

print(f"\n--- Output Gap Response (h = 8) ---")
print(f"  VAR(4): {var4_y[8]: .4f}  [{var4_lo_y[8]: .4f}, {var4_hi_y[8]: .4f}]")
print(f"  LP:     {irf_lp_y[8]: .4f}  [{ci_lo_y[8]: .4f}, {ci_hi_y[8]: .4f}]")
print(f"  VAR(1): {var1_y[8]: .4f}  (misspecified)")

print(f"\n--- Inflation Response (h = 8) ---")
print(f"  VAR(4): {var4_pi[8]: .4f}  [{var4_lo_pi[8]: .4f}, {var4_hi_pi[8]: .4f}]")
print(f"  LP:     {irf_lp_pi[8]: .4f}  [{ci_lo_pi[8]: .4f}, {ci_hi_pi[8]: .4f}]")
print(f"  VAR(1): {var1_pi[8]: .4f}  (misspecified)")

print("\n" + "=" * 60)
